chia/caip2.md

```diff
@@ -0,0 +1,148 @@
+---
+namespace-identifier: chia-caip2
+title: Chia Namespace - Chains
+author: Jeff Cruikshank (@paninaro)
+status: Draft
+type: Standard
+created: 2023-05-24
+requires: CAIP-2
+---
+
+# CAIP-2
+
+*For context, see the [CAIP-2][] specification.*
+
+## Abstract
+
+In CAIP-2 a general blockchain identification scheme is defined. This is the
+implementation of CAIP-2 for the Chia blockchain.
+
+### Chia Namespace
+
+The namespace "chia" refers to the Chia open-source blockchain platform.
+
+## Rationale
+
+Chia maintains multiple blockchains: a production blockchain (mainnet) and one
+or more test blockchains (testnet). Each blockchain is identified by its unique
+genesis challenge. The genesis challenge is a hex-encoded string representing a
+256-bit value that serves as the previous block hash of the first block in the
+blockchain.
+
+An identifier for a Chia blockchain consists of the namespace prefix "chia:"
+followed by a truncated form of the blockchain's genesis challenge.
+
+To conform to the CAIP-2 reference spec, references to Chia's genesis
+challenges are truncated to the first 32 hex characters of the 256-bit value.
+
+Example:
+
+Mainnet's full genesis challenge:
+`ccd5bb71183532bff220ba46c268991a3ff07eb358e8255a65c30a2dce0e5fbb`
+
+Truncation of the full genesis challenge:
+`trunc("ccd5bb71183532bff220ba46c268991a3ff07eb358e8255a65c30a2dce0e5fbb", 32)`  
+`  --> "ccd5bb71183532bff220ba46c268991a"`
+
+CAIP-2 Chia blockchain reference:
+`"chia:ccd5bb71183532bff220ba46c268991a"`
+
+### Resolution Method
+
+To resolve a blockchain reference for the Chia namespace, make a JSON RPC
+request for block 0 using the `get_block_record_by_height` endpoint documented
+at [Full Node RPC Endpoints][]. The value of the `prev_hash` property will
+contain the genesis challenge for the blockchain.
+
+Example:
+
+```
+# Request
+$ curl -X POST
+       --insecure
+       --cert ~/.chia/mainnet/config/ssl/full_node/private_full_node.crt
+       --key ~/.chia/mainnet/config/ssl/full_node/private_full_node.key
+       -H "Accept: application/json"
+       -H "Content-Type: application/json"
+       https://localhost:8555/get_block_record_by_height
```

Comment:
Is there a testnet fullnode or other public RPC endpoint that can be used to test this without running a fullnode locally? If not, could we link to instructions on setting one up? We generally prefer the CLI examples to be runnable by a casual observer or someone still evaluating whether or not to add Chia to a multi-chain project, so making it dead simple to test/grok/play allows this specification to double as an open invitation to curious researchers to start playing with the RPC themselves...

Reply:
Possible alternative: are there public block explorers with RPC endpoints?

Reply:
Thanks for the feedback @bumblefudge. Unfortunately it's not possible to issue an RPC request to a random public full node without writing a fair amount of code. I did add a couple of CLI examples for retrieving the genesis challenge using a public Chia blockchain explorer (Spacescan). I also included a link to installation/setup instructions for running a local full node.

Reply:
@paninaro, I just saw your statement and am truly heartbroken. FireAcademy.io has been running full nodes for well over a year. You can sign up without a credit card, and our free tier includes enough credits that some apps never need to upgrade. Once you generate an API key (for free), calling full node RPC endpoints is as easy as:

```sh
curl -X POST https://kraken.fireacademy.io/[api-key]/leaflet/get_block_record_by_height -H "Content-Type: application/json" -d '{"height": 1337}'
```
Want testnet10 RPC calls? Just switch 'leaflet' to 'leaflet-testnet10':

```sh
curl -X POST https://kraken.fireacademy.io/[api-key]/leaflet-testnet10/get_block_record_by_height -H "Content-Type: application/json" -d '{"height": 1337}'
```

Reply:
I don't have any philosophical objection to updating the spec to include a fireacademy.io example (sorry for slight @Yakuhito! :)), but as a matter of process, this PR has been sitting around for a few weeks with 1/2 needed approvals. I don't know what the process is with the ChainAgnostic folks, but I'm feeling twitchy about making a change that could reset the clock by another several weeks...

@bumblefudge what would you suggest?

Reply:
There is also Mojonode https://api.mojonode.com/docs

Reply:
@denisu Mojonode doesn't support testnet10 (currently). I had pinged trepca about this.

Reply:
don't worry about the timing, i'm fine with it as long as the most current, useful, and complete information is there-- remember, your target audience is people who want to hit the ground running quickly and check their work, without necessarily doing a whole developer onboarding!  ideally, someone equipped with only this document could "double check" their work with CLI tools, so feel free to provide multiple ways of doing that.

Reply:
Thanks @bumblefudge. I've updated the PR to include examples using FireAcademy.